packages/core/src/memory/memory.ts

```diff
@@ -0,0 +1,243 @@
+   * only return messages that are within context window of the LLM
+   * @param llm - To fit the result messages to the context window of the LLM. If not provided, the default token limit will be used.
+   * @param transientMessages - Optional transient messages to include.
+   * @returns The messages from the memory, optionally including transient messages.
+   */
+  async getLLM(
+    llm?: LLM,
+    transientMessages?: ChatMessage[],
+  ): Promise<ChatMessage[]> {
+    const messages = [...this.messages, ...(transientMessages || [])];
+    const contextWindow = llm?.metadata.contextWindow;
+    const tokenLimit = contextWindow
+      ? Math.ceil(contextWindow * DEFAULT_TOKEN_LIMIT_RATIO)
+      : this.tokenLimit;
+
+    return this.applyTokenLimit(messages, tokenLimit);
+  }
+
+  async getMessagesWithLimit(tokenLimit: number): Promise<ChatMessage[]> {
+    return this.applyTokenLimit(this.messages, tokenLimit);
+  }
+
+  async clear(): Promise<void> {
+    this.messages = [];
+  }
+
+  /**
+   * Creates a snapshot of the current memory state
+   * @returns A JSON-serializable object containing the memory state
+   */
+  snapshot(): string {
+    return JSON.stringify({
+      messages: this.messages,
+      tokenLimit: this.tokenLimit,
+    });
+  }
+
+  /**
+   * Creates a new Memory instance from a snapshot
+   * @param snapshot The snapshot to load from
+   * @returns A new Memory instance with the snapshot data
+   */
+  static loadMemory(snapshot: string): Memory {
+    const { messages, tokenLimit } = JSON.parse(snapshot);
+    const memory = new Memory(messages, tokenLimit);
+    return memory;
+  }
+
+  private applyTokenLimit(
+    messages: ChatMessage[],
+    tokenLimit: number,
+  ): ChatMessage[] {
+    if (messages.length === 0) {
+      return [];
+    }
+
+    let tokenCount = 0;
+    const result: ChatMessage[] = [];
+
+    // Process messages in reverse order (keep most recent)
+    for (let i = messages.length - 1; i >= 0; i--) {
+      const message = messages[i];
+      if (!message) continue;
+
+      const messageTokens = this.countMessagesToken([message]);
+
+      if (tokenCount + messageTokens <= tokenLimit) {
+        result.unshift(message);
+        tokenCount += messageTokens;
+      } else {
+        // If we can't fit any more messages, break
+        // But always try to include at least the most recent message
+        if (result.length === 0 && messageTokens <= tokenLimit) {
+          result.push(message);
+        }
+        break;
+      }
+    }
+
+    return result;
+  }
+
+  private countMessagesToken(messages: ChatMessage[]): number {
+    if (messages.length === 0) {
+      return 0;
+    }
+    const tokenizer = Settings.tokenizer;
+    const str = messages.map((m) => extractText(m.content)).join(" ");
+    return tokenizer.encode(str).length;
+  }
+
+  // ========================================
+  // Backward compatibility methods with BaseMemory interface
```

Comment:
doesn't make a lot of sense - we either keep the BaseMemory or not.

I would say we keep it and just use the existing methods